Remove, postpone, complete Task, restructuring.

RodionShyshkin opened the pull request:
Memory model classes were restructured, remove, postpone and completeTask methods were added. 

Documentation and tests were updated.

RodionShyshkin committed:
Restructured and so on

RodionShyshkin committed:
Added postpone & completeTask

ValeriiSudakov commented on src/API/TaskService.cpp:
```diff
@@ -4,74 +4,62 @@
-  return task_service_storage_.GetTasksForToday();
+  auto todayTasks = task_service_storage_.GetTaskView().GetTodayTasks();
+  std::vector<TaskDTO> searchResult;
+  for(auto task : todayTasks) {
```
it's better to use const auto&

ValeriiSudakov commented on src/MemoryModel/Storage/FullStorage.cpp:
```diff
@@ -0,0 +1,74 @@
+  auto newtask = TaskEntity::createTask(task, newid.value());
+  auto task_ptr = std::make_shared<TaskEntity>(newtask);
+
+  if(!task_storage_.PushTask(task_ptr)) return OperationResult::create(ErrorCode::WRONG_TASK_ID);
```
PushTask.. you should to rename

ValeriiSudakov commented on src/MemoryModel/Storage/TaskView.cpp:
```diff
@@ -0,0 +1,106 @@
+
+template <typename T>
+bool removeFromMap(const TaskID& id, std::multimap<T, std::pair<TaskID, std::weak_ptr<TaskEntity>>> map) {
+  for(auto it = map.begin(); it != map.end(); ++it) {
```
why dont u use const auto& : map ?

ValeriiSudakov commented on src/MemoryModel/Storage/TaskView.cpp:
```diff
@@ -0,0 +1,106 @@
+bool TaskView::RemoveTask(const TaskID& id) {
+  bool flag = false;
+
+  if(!removeFromMap(id, priority_sorted_)) return false;
+  if(!removeFromMap(id, name_sorted_)) return false;
+  if(!removeFromMap(id, label_sorted_)) return false;
+  if(!removeFromMap(id, date_sorted_)) return false;
```
If 1 of remove returned false - other will not be removed...

ValeriiSudakov commented on src/MemoryModel/Storage/TaskView.cpp:
```diff
@@ -0,0 +1,106 @@
+
+std::vector<TaskEntity> TaskView::GetWeekTasks() {
+  std::vector<TaskEntity> searchResult;
+  for(auto [date, task] : date_sorted_) {
+    if(Date::CheckWeek(date)) searchResult.push_back(*(task.second.lock()));
+  }
+  return searchResult;
+}
+
+std::vector<TaskEntity> TaskView::GetTasksByPriority(const Priority &priority) {
+  auto first = priority_sorted_.find(priority);
+  std::vector<TaskEntity> searchResult;
+
+  if(first == priority_sorted_.end()) return searchResult;
+  else {
+    for(auto it = first; it != priority_sorted_.end(); ++it) {
```
Why do u use different iterators in same loops?

ValeriiSudakov commented on src/MemoryModel/Task/TaskEntity.cpp:
```diff
@@ -0,0 +1,49 @@
+  return TaskEntity(task, id, parent);
+}
+
+TaskID TaskEntity::GetID() const { return user_id_; }
+std::string TaskEntity::GetName() const { return task_.GetName(); }
+std::string TaskEntity::GetLabel() const { return task_.GetLabel(); }
+Priority TaskEntity::GetPriority() const { return task_.GetPriority(); }
+Date TaskEntity::GetDueTime() const { return task_.GetDate(); }
+bool TaskEntity::GetStatus() const { return status_; }
+Task TaskEntity::GetTask() const { return task_; }
+std::map<TaskID, std::shared_ptr<TaskEntity>> TaskEntity::GetSubtasks() const { return subtasks_; }
+TaskID TaskEntity::GetParentID() const { return parent_id_; }
```
Add spaces between methods..............

ValeriiSudakov commented on src/MemoryModel/Task/TaskEntity.h:
```diff
@@ -0,0 +1,93 @@
+  TaskID                                                  user_id_;
+  bool                                                    status_;
+  Task                                                    task_;
+  std::map<TaskID, std::shared_ptr<TaskEntity>>           subtasks_;
```
R u sure, that shared ptr is necessary here? mb weak ptr?

RodionShyshkin committed:
changed data structure in TaskView
shared_ptr was replaced by weak_ptr in TaskEntity
loops by iterators were replaced by range based ones

RodionShyshkin committed:
Added sorting by priority

RodionShyshkin committed:
Added documentation and tests were updated.
